File: 05/c++/part2.cpp
#include <iostream>
#include <string>
#include <sstream>
#include <fstream>

#define INPUT "../data/input"

/**
 * Read a file entirely into a stringstream
 * 
 * @param The name of the file to read
 * @return The contents of the file
 **/
std::stringstream readFile(const std::string& filename = INPUT) {
    std::ifstream fileStream(filename);
    std::stringstream returnVal;
    
    returnVal << fileStream.rdbuf();

    return returnVal;
}

bool rule1(const std::string& string) {
    for (decltype(string.size()) index = 0; index < string.size() - 3; ++index) {
        for (decltype(string.size()) index2 = index + 2; index2 < string.size() - 1; ++index2) {
            if ((string[index2] == string[index]) && 
                (string[index2 + 1] == string[index + 1])) {
                return true;
            }
        }
    }

    return false;
}

bool rule2(const std::string& string) {
    for (decltype(string.size()) index = 0; index < string.size() - 2; ++index) {
        if (string[index] == string[index + 2]) {
            return true;
        }
    }

    return false;
}

int main() {
    std::stringstream contents = readFile();
    std::string line;
    uint64_t numNice = 0;

    // For each line in file
    while (std::getline(contents, line, '\n')) {
        // Check for double pairs
        if (!rule1(line)) {
            continue;
        }

        // Check for xyx combinations
        if (!rule2(line)) {
            continue;
        }

        ++numNice;
    }

    std::cout << "Number of nice strings : " << numNice << std::endl;

    return 0;
}

File: 05/python/part2.py
#!/usr/bin/env python

def rule1(string):
    for index in range(0, len(string) - 3):
        subString = string[index:index + 2]

        if (subString in string[index + 2:]):
            return True

    return False


def rule2(string):
    for index in range(0, len(string) - 2):
        if (string[index] == string[index + 2]):
            return True

    return False


inputFile = open("../data/input", "r")

numNiceStrings = 0

for string in inputFile.readlines():
    string = string.strip()

    if (not rule1(string)):
        continue

    if (not rule2(string)):
        continue

    numNiceStrings += 1

print("Number of nice strings:", numNiceStrings)
